pkg/compact/compact.go

```diff
@@ -447,12 +447,15 @@ func (cg *Group) Resolution() int64 {
 
 // Compact plans and runs a single compaction against the group. The compacted result
 // is uploaded into the bucket the blocks were retrieved from.
-func (cg *Group) Compact(ctx context.Context, dir string, comp tsdb.Compactor) (bool, ulid.ULID, error) {
+func (cg *Group) Compact(ctx context.Context, dir string, comp tsdb.Compactor) (shouldRerun bool, compID ulid.ULID, rerr error) {
 	cg.compactionRunsStarted.Inc()
 
 	subDir := filepath.Join(dir, cg.Key())
 
 	defer func() {
+		if rerr != nil {
```

Comment:
Is it correct doing it for any error?

Reply:
I think so:

```go
	if err := os.RemoveAll(subDir); err != nil {
		return false, ulid.ULID{}, errors.Wrap(err, "clean compaction group dir")
	}
```
Most likely the same operation will fail again and the deletion of it in `defer` is best-effort either way.


```go
	if err := os.MkdirAll(subDir, 0777); err != nil {
		return false, ulid.ULID{}, errors.Wrap(err, "create compaction group dir")
	}
```

This is not a `HaltError` and then in such case, the caller will clean up the parent directory in a best-effort manner. Is there something, in particular, that doesn't look correct to you?

Reply:
This function can return also if `cg.compact()` returns error. There are many reasons why `cg.compact()` could return error, and it can also return an `HalfError`. The `os.RemoveAll()` done in this `defer` is used to cleanup the blocks downloaded (or created) by `cg.compact()`. For this reason, I'm wondering what we want to achieve with this change. If it's all about preserving the disk state on halt error, then I think we could also check if the HaltError here as well.